### ISSUE TITLE
Add hookable weights

### ISSUE THREAD

**Issue by daemon** [pytorch/pytorch#5790]
It would be nice to have hookable weights, where per-layer parameters can be operated upon and determined dynamically during each forward pass. The following is a short list of techniques that could benefit from such a framework:

- Quantization[[1]](https://arxiv.org/abs/1609.07061)
- Pruning[[2]](https://arxiv.org/abs/1510.00149)[[3]](https://arxiv.org/abs/1707.06168)[[4]](https://arxiv.org/abs1802.00124)
- Stochastic gradient estimation[[5]](https://arxiv.org/abs/1711.00123)[[6]](https://arxiv.org/abs/1703.07370)
- DropConnect[[7]](https://cs.nyu.edu/~wanli/dropc/)
- Regularization[[8]](https://openreview.net/forum?id=H1Y8hhg0b)
- Manifold Tangent Classifier[[9]](https://papers.nips.cc/paper/4409-the-manifold-tangent-classifier.pdf)

I've begun work on [such a framework](https://github.com/castorini/candle), but the code is research-oriented and not production ready. I think the community would benefit from having official support for hookable weights.

I think it would also be interesting to have a more extensible backprop framework, in the sense of allowing for computation of user-defined quantities. For example, the Hessian diagonals can be computed efficiently[[10]](http://yann.lecun.com/exdb/publis/pdf/lecun-90b.pdf) using a backprop-like approach.

Thanks for reading.

**Comment by zou3519** [comment]
Could you provide an example of what you mean by hookable weights?

**Comment by daemon** [comment]
Yep. The idea is to pass some weights _w_ through a user-specified function _g(w)_ for each forward pass, before the layer operates on the input. _g(w)_ is then used for the weights instead of _w_ for that layer. _g_ would of course be the identity function in the normal case. Here are a few practical examples:

1. Pruning
We would like to zero out weights according to some criterion. Yes--we can zero the weights with the current implementation, but it's more elegant to have something like the following:

```python
linear = nn.Linear(1024, 1024)
conv = nn.Conv2d(3, 64, 3)
...
def magnitude_prune(weight, max_magnitude=..., pct=...):
    ...
    return prune_output
linear.hook_weight(magnitude_prune, max_magnitude=0.01, pct=0.1)
conv.hook_weight(magnitude_prune, max_magnitude=0.1, pct=0.1)
```

This could represent pruning 0.1% of the weights (at every forward pass) that are less than 0.1. Beyond this simple example, there are other methods that necessarily involve hooking weights. Consider [this paper](https://openreview.net/forum?id=H1Y8hhg0b):

```
function g(w):
    z <- z ~ p(z | phi)
    w_new <- z * w
    return w_new

layer.hook_weight(g)
```

where `z ~ p(z | phi)` is the (continuous) hard concrete distribution, `w` are the original weights, and `phi` is a learnable parameter. During training, `dL / dphi` is computed using backprop and used to update `phi`. Since `phi` is a parameter, this suggests that `g` can also take on the form of an `nn.Module`.

2. Quantization
We would like to quantize some weights _w_. Consider the case of [1-bit quantization](https://arxiv.org/abs/1609.07061):

```python
class BinarizeFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        """Outputs +/- 1 depending on magnitude"""
        return x.clamp(0, 1).ceil() * 2 - 1

    @staticmethod
    def backward(ctx, grad_output):
        return grad_output # straight-through estimator

binarize_function = BinarizeFunction.apply

def binarize(w):
    w.clamp_(0, 1)
    return binarize_function(w)

layer.hook_weight(binarize)
```

This isn't completely correct; it's just an illustrative example. For more concrete examples, please see[[1]](https://github.com/castorini/candle/blob/master/candle/proxy.py)[[2]](https://github.com/castorini/candle/blob/master/candle/prune.py)[[3]](https://github.com/castorini/candle/blob/master/candle/regularize.py).

**Comment by rosinality** [comment]
What about the approach of the weight_norm implementation? I think it solves similar problem. Certainly it is more verbose than the simple hooking functions in your example, though.

**Comment by apaszke** [comment]
If `hook_weight` runs at every forward, why can't you just use `register_forward_hook` and access `module.weight` inside? It shouldn't be any more complicated than the usage of `hook_weight`

**Comment by fmassa** [comment]
I thought in those cases we would encourage the use of the functional interface for writing the computation graph? I think assigning to a module weight in the `forward_hook` might not be permitted actually.

What about something like
```python
class MyModel(nn.Module):
    def __init__(self):
        self.conv1 = nn.Conv2d(1, 1, 1)
        self.conv2 = nn.Conv2d(1, 1, 1)
        self.conv3 = nn.Conv2d(1, 1, 1)

    def _my_crazy_op(self, x, layer):
        weight = layer.weight
        bias = layer.bias
        weight = a_crazy_func(weight)
        bias = another_crazy_func(bias)
        return F.conv2d(x, weight, bias, stride, ...)

    def forward(self, x):
        x = self.conv1(x)
        x = self._my_crazy_op(x, self.conv2)
        x = self._my_crazy_op(x, self.conv3)
        return x
```

**Comment by daemon** [comment]
@apaszke We could register a `forward_pre_hook` and modify the weights there, but it'd be better to distinguish between the actual module weights and what's provided by our custom ops, which might have parameters themselves. I guess we could then revert to the original weights in a `forward_hook`, but that's not as concise as having explicit weight hooks/providers.

@fmassa Yes, that would work, but then we have to worry about bias, stride, etc. Actually, that amounts to writing a weight hooking layer:
```python
class HookedConv2d(nn.Module):
    def __init__(self, weight_provider, registry=None):
        super().__init__()
        self.weight_provider = weight_provider

    def hook_weight(self, weight_hook, **kwargs):
        self.weight_provider = weight_hook(self.weight_provider, **kwargs)

    def forward(self, x):
        weights = self.weight_provider()
        return F.conv2d(x, *weights, **...)
```

**Comment by daemon** [comment]
Hey, just bumping this issue.. I really think a lot of people could benefit from this. For example, the popular [AWD-LSTM ](https://github.com/salesforce/awd-lstm-lm) (out of necessity) uses a janky hook for LSTM weights literally breaks with every minor version release, since there's no official support for hookable weights.

**Comment by soumith** [comment]
hey, I thought about this a bit.

I dont think hookable weights are the right solution, because we already do provide stuff that's very close.
FIrst, we provide [model.named_children](https://pytorch.org/docs/stable/nn.html?highlight=named_children#torch.nn.Module.named_children), `named_modules` and `named_parameters` via which you can comfortably do something like:

```python
prune_model_(model)
out = model(inp)
...
```

where `prune_model` loops over the layers, and for each instance (via the variable name or via `isinstance` do something reasonable).

As mentioned above, you give a scenario that your weight pruning logic might have Parameters itself -- well in that case, you are forced to either write separate layers, or register those Parameters to either the existing `model` or to a separate prune learning model -- any of the above solutions should work right?

If you're not convinced, I'd like to discuss this further and get this thread to a closure.

On  the topic of pruning, @mickypaganini has opened https://github.com/pytorch/pytorch/issues/20402 which I'm more inclined to accept under `torch.nn.utils`, so I'd like to get this topic further along in discussion as well.

Again, sorry for the delay.

**Comment by daemon** [comment]
Thanks for your input.
> where prune_model loops over the layers, and for each instance (via the variable name or via isinstance do something reasonable).

> @mickypaganini has opened #20402 which I'm more inclined to accept under torch.nn.utils

Those solutions are reasonable for parameterless pruning.

> you give a scenario that your weight pruning logic might have Parameters itself -- well in that case, you are forced to either write separate layers, or register those Parameters to either the existing model or to a separate prune learning model

For most problems, it's true that @apaszke's and @fmassa's solutions work, e.g., either implementing the logic in a forward hook or a module with `functional` calls. The main problem is elegance and modularity. The issue with @apaszke's suggestion is that the weights need to be explicitly set. Using forward hooks don't play well with multiple weight hooks either, which forces the user to write more boilerplate. For example, if I have both a quantization and a pruning component, I would need to write some code to marry the two to the same group of weights. It's not hard, it just requires boilerplate.

On the other hand, weight hooks conveniently handle these two problems: first, we don't need to explicitly set the weights; second, we can chain together multiple hooks. Both of these have clear benefits for writing more modular code:

```python
module = MyModule()
module.register_hook('weight', QuantizeHook(bits=8))
module.register_hook('weight', LearnablePruningHook())
module.register_hook('weight', DropConnectHook())
module.register_hook('scale', SomeOtherHook())
```

In the forward hook solution, we have to manually set the weights, as well as "coordinate" the right set of weights between the multiple hooks. Neither manually setting nor coordinating the weights is a terrible solution; it just requires more boilerplate. In the example above, all `QuantizeHook` has to contain is the quantization logic. It doesn't need to worry about setting the weights, coordinating weights, or vice versa. That's all the user would have to write for modular code. I guess this is analogous to Python's @property; by decoupling the weight getter from the attribute, we enable more complex, modularized operations over the weights.

As for @fmassa's suggestion, a similar argument can be made. This time, we also have to worry about the (hyper)parameters of the target functional, which shouldn't belong in, say, `QuantizeHook`. Furthermore, that suggestion doesn't work with LSTMs and GRUs.

Again, I think the alternative solutions are fine. I understand this thread isn't top priority for PyTorch -- I just think hookable weights would improve over the existing solutions, mostly with respect to readability and modularity. It also follows the style of the existing forward hooking mechanisms.

**Comment by soumith** [comment]
the above sample does look tempting.
One problem that affects hooks -- but especially affects this style of code (because it's doing considerable non-trivial work) is that hooks aren't serializable -- storing so much state about the mechanism of the model in something like this seems laden with side-effects (esp. across serialization-deserialization).
But I think you are building a case for hooks (I am still not convinced though ;-) ).
I have seen something similar, but gotten in a different way, with weight_norm: https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/weight_norm.py , but not everyone has liked it.

**Comment by daemon** [comment]
> One problem that affects hooks -- but especially affects this style of code (because it's doing considerable non-trivial work) is that hooks aren't serializable

Hmm, I guess `model.(named_)parameters()` would additionally have to return the parameters of the weight hooks. I think the same holds for forward hooks as well. I also see your point about deserialization. If I'm unmistaken, the user must provide the correct "namespace" with the appropriate classes at deserialization, which already happens in some of the code I see in practice.

I suppose if users follow the recommended PyTorch serialization-deserialization route, it should be fine:

> When saving a model for inference, it is only necessary to save the trained model’s learned parameters. Saving the model’s state_dict with the torch.save() function will give you the most flexibility for restoring the model later, which is why it is the recommended method for saving models.

```python
# Serialization
module = MyModule()
module.register_forward_hook(mean_activation_prune)
module.register_weight_hook('weight_name', QuantizeHook(bits=8))
sd = module.state_dict()
'hooks.weight.weight_name.quantize' in sd # True
'hooks.forward.mean_activation_prune' in sd # True
torch.save(sd, 'state_dict.pt')

# Deserialization
sd = torch.load('state_dict.pt')
module = MyModule()
module.register_forward_hook(mean_activation_prune)
module.register_weight_hook('weight_name', QuantizeHook(bits=8))
model.load_state_dict(sd)
```

**Comment by chenbohua3** [comment]
Any updates? Another problem may be the gradient of the weight. In quantization aware training, we may want to modify the conv weight manually and make the `scale` and `zero point` in the backward graph to make them updated end-to-end.

Here is a sample:
```

class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.layer = nn.Conv2d(1, 2, 3, bias=False)
        self.scale = torch.nn.Parameter(torch.tensor([0.5]))

    def forward(self, x):
        new_weight = self.layer.weight
        self.scale = some_operation(self.scale)
        new_weight = new_weight * self.sclae
        self.layer.weight = new_weight 
        x = self.layer(x)
        return x
```

Since the final new_weight is of type `torch.Tensor`, we can not directly assign it to  `self.layer.weight` . If we construct a new `torch.nn.Parameter(new_weight)` and assign it to `self.layer.weight`, then the scale will not appear in the backward graph.

**Comment by zou3519** [comment]
Clearing the assignees. This sounds a lot like https://github.com/pytorch/pytorch/issues/49171

**Comment by albanD** [comment]
The original question was mostly about reparametrization I think that was added in https://github.com/pytorch/pytorch/issues/28937

Does that satisfy your use case?
